Convert TravisCI to GitHub Actions

VadimSchmitz opened the pull request:
The following work is done in this pull request.

-TravisCI has been replaced with Composer scripts and a GitHub Actions workflow
-Static analysis PHPmd & PHPcs has been added.
-Updated README and .gitignore

VadimSchmitz committed:
Update .gitignore

The files added to the gitignore are generated by the tests. therefore they wont need to be added to the commit.

VadimSchmitz committed:
Update readme

VadimSchmitz committed:
Convert Travisci to GithubActions

In this commit the following work has been done:
- Add a GitHub Actions workflow for test-integration
- Add docker-compose to run the tests on a docker container
- Move phpunit and security tests to composer.json
- Remove travis.yml file

squash

VadimSchmitz committed:
Add static analysis: PHPmd & PHPcs

Add static code analysers PHPmd & PHPcs for a more consistent code quality.
Currently both tests are allowed to fail due to it being legacy code.

VadimSchmitz committed:
Seperate static-analysis from other tests.

Currently the static analysis tests give alot of warnings due to the code being legacy code. To keep the Github Action output useful the static analysis has been split into its own worflow step.

VadimSchmitz committed:
Merge branch 'develop' into feature/travis-to-gha

VadimSchmitz committed:
Delete .travis.yml

VadimSchmitz committed:
Fix GHA workflow

Fix typo and workflow call
Move back files that moved after the merge

MKodde commented on .github/workflows/test-integration.yml:
```diff
@@ -0,0 +1,31 @@
+      - name: Install dependencies
+        run: cd ci/docker && docker-compose exec -T tiqr-server-libphp bash -lc 'composer install'
+      - name: Run Static analysis
+        run: cd ci/docker && docker-compose exec -T tiqr-server-libphp bash -lc 'composer static-analysis'
+      - name: Run test scripts
+        run: cd ci/docker && docker-compose exec -T tiqr-server-libphp bash -lc 'composer phpunit && composer security-tests'
```
Why not create a `test` composer script, and run the required tests from there? That's how we do it in the other projects right? I'm also fine with keeping it this way. It's easy enough to restructure when the need arises.

VadimSchmitz replied:
As explained in person, Currently the Static analysis tests are allowed to fail due to the code being legacy code. However the Static analysis gives a lot of style warnings and errors. Therefore i felt that it was a good idea to split these tests up otherwise you would not be able to read the outputs of the other tests easily.

MKodde replied:
Good suggestion!

MKodde commented on .github/workflows/test-integration.yml:
```diff
@@ -0,0 +1,31 @@
+        run: cd ci/docker && docker-compose exec -T tiqr-server-libphp bash -lc 'composer phpunit && composer security-tests'
+      - name: Output log files on failure
+        if: failure()
+        run: cd ci/docker && docker-compose exec -T tiqr-server-libphp -c 'cat var/log/webtest.log'
```
Maybe cat the syslog here? var/log/webtest.log will certainly not exist.

VadimSchmitz replied:
This will indeed not work, it has been changed.

MKodde commented on README.md:
```diff
@@ -1,7 +1,9 @@
+`Master` ![test-integration workflow](https://github.com/SURFnet/tiqr-server-libphp/actions/workflows/test-integration.yml/badge.svg?branch=master)
+[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/SURFnet/tiqr-server-libphp/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/SURFnet/tiqr-server-libphp/?branch=master)
+<br>
+`Develop` ![example workflow](https://github.com/SURFnet/tiqr-server-libphp/actions/workflows/test-integration.yml/badge.svg?branch=develop)
```
`example workflow` should be test-integration workflow?

VadimSchmitz replied:
fixed

MKodde commented on ci/docker/docker-compose.yml:
```diff
@@ -0,0 +1,10 @@
+    volumes:
+      - ../../:/var/www
+    environment:
+      APP_ENV: test
```
Given this is not a Symfony project. The APP_ENV wil probably not be evaluated.

VadimSchmitz replied:
It has been cleaned up

MKodde commented on ci/qa/phpunit:
```diff
@@ -0,0 +1,6 @@
+cd $(dirname $0)/../../
+
+printf "Running PHPUnit\n"
+XDEBUG_MODE=coverage vendor/bin/phpunit --configuration=ci/qa/qa-config/phpunit.xml
```
Please disable code coverage on the phpunit run. I decided on this with Pieter in one of the PR's. We'll use scrutenizer to show code coverage reports.

VadimSchmitz replied:
The Xdebug code coverage has been turned off for now.

MKodde commented on ci/qa/phpunit:
```diff
@@ -0,0 +1,6 @@
+
+cd $(dirname $0)/../../
+
+printf "Running PHPUnit\n"
```
:bow:

MKodde commented on ci/qa/qa-config/phpcs.xml:
```diff
@@ -0,0 +1,25 @@
+    <arg name="colors"/>
+    <arg name="extensions" value="php"/>
+
+    <rule ref="PSR2"/>
```
The code is defenitely not complient with the PSR2 code style. The ZF code style will be a close match. Maybe try to run the CS against those rules?

`https://github.com/zendframework/zend-coding-standard/blob/master/README.md`

VadimSchmitz replied:
Changed the standard to use the Zend standard.

MKodde commented on composer.json:
```diff
@@ -15,6 +15,18 @@
+        "squizlabs/php_codesniffer": "^3.6"
+    },
+    "scripts": {
+        "test": [
+            "@static-analysis",
+            "@phpunit",
+            "@security-tests"
+        ],
```
Please run the `composer test` script directly from your Github Actions workflow

VadimSchmitz replied:
see https://github.com/SURFnet/tiqr-server-libphp/pull/34#discussion_r844790811

MKodde commented on ci/qa/static-analysis:
```diff
@@ -0,0 +1,15 @@
+#phpcs static analysis
+./vendor/bin/phpcs --standard=ci/qa/qa-config/phpcs.xml --report=full --warning-severity=0 ./library
+
+exit 0
```
Looking good!

MKodde reviewed (changes requested):
A very nice first setup! See some feedback below.

VadimSchmitz committed:
Fix feedback points

Things fixed and/improved in this commit:
-Cleanup test-integration.yml
-Cleanup REAMDE.md
-Cleanup docker-compose.yml
-PHPunit turn off Xdebug
-Use Zend coding standard instead of PSR2 in phpcs

MKodde commented on ci/qa/qa-config/phpcs.xml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ruleset name="Zend Framework Coding Standard">
+
+    <rule ref="../../../vendor/zendframework/zend-coding-standard/ruleset.xml"/>
+
+    <!-- Paths to check -->
+    <file>bin/</file>
+    <file>config/</file>
+    <file>public/</file>
+    <file>src/</file>
+    <file>dev/</file>
+
+        <!-- Lines can be a little bit longer before they break the build -->
+    <rule ref="Generic.Files.LineLength">
+        <properties>
+            <property name="lineLimit" value="120"/>
+            <property name="absoluteLineLimit" value="150"/>
+        </properties>
+    </rule>
+</ruleset>
```
:metal:

MKodde reviewed (approved):
Grand!
